### ISSUE TITLE
Consider introducing Views.isDefined()

### ISSUE THREAD

**Issue by stelfrich** [imglib/imglib2#171]
Introducing `Views.isDefined(RandomAccessibleInterval, Interval)` would enable Ops to check if a `RandomAccessibleInterval` needs an extension for a computation to succeed.

Currently, filter-like Ops (that might access pixels outside of the `Interval`) can be provided an optional `OutOfBoundsStrategy` for convenience. If none is provided, we apply a default one without checking if that's actually necessary. This overwrites an already existing `OutOfBoundsStrategy`. While you can always set the `oobf` to `null` when executing an Op on an already extended image, we want to be more intelligent about when to extend an image.

Work happening on https://github.com/imglib/imglib2/tree/defined-bounds

**See also**
- https://github.com/stelfrich/imglib2/commits/random-access-defined-bounds-no-interfaces
- [AbstractNeighborhoodBasedFilter.java#L53-L55](https://github.com/imagej/imagej-ops/blob/master/src/main/java/net/imagej/ops/filter/AbstractNeighborhoodBasedFilter.java#L53-L55)

/cc @tpietzsch

**Comment by ctrueden** [comment]
See also #167.

**Comment by stelfrich** [comment]
Sorry for the delay, @tpietzsch! I have added some tests that are currently failing (as expected) based on our discussion at the Learnathon.

**Comment by hanslovsky** [comment]
Wouldn't that be equivalent to [`Intervals.contains( containing, contained )`](https://github.com/imglib/imglib2/blob/master/src/main/java/net/imglib2/util/Intervals.java#L483-L494)?

**Comment by ctrueden** [comment]
Firstly, note that there is some [quasi-related discussion on Gitter](https://gitter.im/imglib/imglib2?at=5acb7b226d7e07082bd5cab2).

@stelfrich Here is another way we could go. For ops that rely on "going out of bounds" of the target interval:

* Make a low-level op that takes two arguments, a `RandomAccessible` and an `Interval`. The caller is responsible for ensuring the RA is defined everywhere it needs to be.
* Make a higher-level op that takes one argument, a `RandomAccessibleInterval` and an optional `OutOfBoundsFactory`. It then takes care of converting the RAI to an RA using the OOBF provided (or default OOBF type if none given), and calls the low-level op.

Of course, needing to write two ops is a lot of boilerplate for this common pattern. But maybe we could take care of always generating the high-level op whenever low-level one of that pattern exists. (Similar to how we plan to e.g. always generate a function op for every computer op by leaning on ops in the `create` namespace.)

**Comment by stelfrich** [comment]
> Wouldn't that be equivalent to `Intervals.contains( containing, contained )`?

I don't think so (although I might be mistaken). For the cases that I had in mind, `contained` will always be larger (in every dimensions) than `containing` so that the proposed call will evaluate to `false`. In essence, `contained` is something like `Intervals.expand(containing, kernelSpan)` for filters. Does that make sense, @hanslovsky, or am I adding to the confusion?
Sticking to the filter example (and somewhat connecting to the discussion in Gitter), we would either need a `RA + I` (where we know that it's defined even outside of `I` since it's a `RA`) or some way to check **once** if a `RAI` is not only defined in `I` but also in `I+kernelSpan` (very likely if it has been created by `Views.interval(Views.extend(RAI), I)`.

**Comment by ctrueden** [comment]
@stelfrich Yet another idea that occurs to me: a computer op could use the `Interval` bounds of the _preallocated output_, and accept only an (extended) `RandomAccessible` as the input. One thing that Ops does not make explicit is "how big should the preallocated output be"? And the assumption has always been "same size as the input". But if the input has no interval bounds, then you could pass in whatever you like, I suppose.

The only tricky part then would be auto-converting this computer to a function, so that the output can be automagically allocated. But I think the code could be smart enough to unwrap any input `RandomAccessible` that is actually an extended RAI under the hood, and use the unwrapped `Interval` bounds when creating the output.

**Comment by hanslovsky** [comment]
@stelfrich I might have a wrong understanding of what `Views.isDefined` is supposed to do.
As far as I understand, you would like to know if the containing interval's `RandomAccess` can produce a value at every pixel of the contained interval. Quoting the [`RandomAccessibleInterval` doc](https://github.com/imglib/imglib2/blob/master/src/main/java/net/imglib2/RandomAccessibleInterval.java):
> By convention, a RandomAccessibleInterval represents a function that is
 *defined at all coordinates of the interval*.

So all we can do is to make sure that the containing interval fully contains the contained interval:
```java
RandomAccessibleInterval containing;
Interval contained = // Create contained interval, e.g. Views.expand( someInterval, kernelSize );
Views.isDefined( containing, contained );
// or, equivalent under my presumption:
Intervals.contains( containing, contained );
```

If `Views.isDefined`, on the other hand, is supposed to automagically check if the `RandomAccessibleInterval containing` is actually a cascade of views and at some point in that cascade there is an `OutOfBounds` access, and hopefully `containing.randomAccess()` would use that `OutOfBounds`, I would very much prefer to go with what @ctrueden [proposed](https://github.com/imglib/imglib2/issues/171#issuecomment-379794035) for the reasons stated in the [discussion on Gitter](https://gitter.im/imglib/imglib2?at=5acb7b226d7e07082bd5cab2) that @ctrueden shared in this thread.

**Comment by hanslovsky** [comment]
> So the idea of this issue being filed is that you could have an RAI that is defined outside its boundaries sufficiently far that an Ops algorithm could safely perform e.g. neighborhood-based calculations at the interval boundaries.

Thank you for the clarification.

> I just want to make sure we are all clear here that accessing an RAI outside the bounding Interval has no explicitly defined behavior—we do not promise, e.g., that some exception will be thrown, or that it will not work.

Yes the `RAI` does not explicitly define any out of bounds behavior but that is also the exact reason why I think `RA` should be used instead. Technically, the `RA` does not guarantee access at all (`long`) coordinates but I think that is owed to the hierarchy (`RAI` extends `RA`), which is not 100% accurate if you think of `RA` and `RAI` of functions over domains `Z^n` and `[min, max] x [min, max] x ...`, respectively. At least, however, `RA` parameters indicate that values outside the region of intertested may be queruied.

> I guess it is probably for the best that no algorithm (Ops or otherwise) ever attempt to access an RAI outside its bounding interval. Agreed?

I agree with that very strongly. I would even go as far as considering such practice a bug.

**Comment by ctrueden** [comment]
> I would even go as far as considering such practice a bug.

Then I think we should document that in the `RandomAccessibleInterval` interface explicitly. There should be a statement urging people to never rely on out of bounds access—behavior is formally undefined, and in practice will often/usually fail, and probably in inconsistent ways across implementations.

> Technically, the `RA` does not guarantee access at all

Indeed, which is largely what bothers me about this distinction. For RA it is the caller's responsibility to ensure it is defined in the areas of interest. That same strategy could allow an RAI to be used with some limited out-of-bounds access, when the caller has preexisting knowledge that such access will be in defined ranges. This would be more consistent, and does not seem like a "bug" to me.
